### PR TITLE
Create the baseDataObject with a counter

### PR DESCRIPTION
**Pull request by tyler-cai-microsoft**
Part of [AB#1847](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/1847)

As a part of https://github.com/microsoft/FluidFramework/pull/11928/files#

`DataObjectWithCounter` is a DataObject containing a `SharedCounter` and a way of retrieving that `SharedCounter` via a handle. It generates ops by incrementing the counter.

### CODE REVIEW

**Comment by agarwal-navin** [comment]
Can you move all the different data objects for the tests in a single file? Too many files creates confusion and its harder to read.

File: packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import { SharedCounter } from "@fluidframework/counter";
+
+/**
+ * DataObjectWithCounter increments a SharedCounter as a way of sending ops.
+ *
+ * The SharedCounter is retrieved via handle
+ */
+const counterKey = "counter";
+export class DataObjectWithCounter extends DataObject {
+    private _counterHandle?: IFluidHandle<SharedCounter>;
```
Comment (agarwal-navin) [inline packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts:19]: It is probably better to directly store the counter and expose that. Its cheaper to get the handle from the counter than vice-versa.

File: packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        assert(this._counterHandle !== undefined, `CounterHandle should always be defined!`);
+        return this._counterHandle;
+    }
+
+    protected async initializingFirstTime(props?: any): Promise<void> {
+        this.root.set<IFluidHandle>(counterKey, SharedCounter.create(this.runtime).handle);
+    }
+
+    protected async hasInitialized(): Promise<void> {
+        this._counterHandle = this.root.get<IFluidHandle<SharedCounter>>(counterKey);
```
Comment (agarwal-navin) [inline packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts:35]: You should assert here that handle exists and then set this._counterHandle.

File: packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public async sendOp(_count: number, _random: IRandom) {
+        assert(this.counterHandle !== undefined, "Can't send ops when counter handle isn't set!");
```
Comment (agarwal-navin) [inline packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts:39]: this.counterHandle is never undefined as per the getter signature. If it is, the getter will throw an assert and the code will never reach here anyway.

File: packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        const counter = await this.counterHandle.get();
```
Comment (agarwal-navin) [inline packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts:40]: If you store the counter directly as per my other comment, you can skip this step which is expensive.

File: packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * DataObjectWithCounter increments a SharedCounter as a way of sending ops.
+ *
+ * The SharedCounter is retrieved via handle
+ */
+const counterKey = "counter";
+export class DataObjectWithCounter extends DataObject {
+    private _counterHandle?: IFluidHandle<SharedCounter>;
+    public isRunning: boolean = false;
```
Comment (agarwal-navin) [inline packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts:20]: Why is this needed?

Comment (tyler-cai-microsoft) [reply packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts:20]: Good point, I added it to the sendOp to prevent ops from being sent when the dataobject is stopped.

File: packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * DataObjectWithCounter increments a SharedCounter as a way of sending ops.
+ *
+ * The SharedCounter is retrieved via handle
+ */
+const counterKey = "counter";
+export class DataObjectWithCounter extends DataObject {
+    private counter?: SharedCounter;
+    public isRunning: boolean = false;
+    public static get type(): string {
```
Comment (agarwal-navin) [inline packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts:20]: Can you tell me why is this type needed (I have not looked at how this is currently used in the stress tests)?

**Review by agarwal-navin — APPROVED** [review]

File: packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts (hunk 6 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-gc-sweep-tests/src/dataObjectWithCounter.ts:20]: It's for the registry entry for the factory
